usersync/ejector.go

```diff
@@ -0,0 +1,145 @@
+package usersync
+
+import (
+	"errors"
+	"math"
+	"time"
+)
+
+type Ejector interface {
+	Choose(uids map[string]UIDEntry) (string, error)
+}
+
+type OldestEjector struct {
+	nonPriorityKeys []string
+	PriorityGroups  [][]string
+	EjectPriority   bool
+}
+
+type PriorityBidderEjector struct {
+	PriorityGroups [][]string
+	SyncerKey      string
+	OldestEjector  OldestEjector
+}
+
+// Choose method for oldest ejector will return the oldest uid after determing which set of elements to be choosing from
+func (o *OldestEjector) Choose(uids map[string]UIDEntry) (string, error) {
+	var elements []string
+
+	// Set which elements the ejector will be choosing from
+	if len(o.nonPriorityKeys) > 0 {
+		elements = o.nonPriorityKeys
+	} else if o.EjectPriority {
+		elements = o.PriorityGroups[len(o.PriorityGroups)-1]
+	} else {
+		elements = getNonPriorityKeys(uids, o.PriorityGroups)
+	}
+
+	uidToDelete := getOldestElement(elements, uids)
+
+	return uidToDelete, nil
+}
+
+// Choose method for priority ejector will return the oldest lowest priority element
+func (p *PriorityBidderEjector) Choose(uids map[string]UIDEntry) (string, error) {
+	p.OldestEjector.nonPriorityKeys = getNonPriorityKeys(uids, p.PriorityGroups)
+	if len(p.OldestEjector.nonPriorityKeys) > 0 {
+		p.OldestEjector.EjectPriority = false
+		return p.OldestEjector.Choose(uids)
+	}
+
+	if isSyncerPriority(p.SyncerKey, p.PriorityGroups) {
+		lowestPriorityGroup := p.PriorityGroups[len(p.PriorityGroups)-1]
+
+		if len(lowestPriorityGroup) == 1 {
+			uidToDelete := lowestPriorityGroup[0]
+			p.PriorityGroups = removeElementFromPriorityGroup(p.PriorityGroups, uidToDelete)
+			return uidToDelete, nil
+		} else {
+			p.OldestEjector.EjectPriority = true
+			p.OldestEjector.PriorityGroups = p.PriorityGroups
+			uidToDelete, err := p.OldestEjector.Choose(uids)
+			if err != nil {
+				return "", err
+			}
+			p.PriorityGroups = removeElementFromPriorityGroup(p.PriorityGroups, uidToDelete)
+			return uidToDelete, nil
+		}
+	}
+	return "", errors.New("syncer key " + p.SyncerKey + " is not in priority groups")
+}
+
+// updatePriorityGroup will remove the selected element from the priority groups, and will remove the entire priority group if it's empty
+func removeElementFromPriorityGroup(priorityGroups [][]string, oldestElement string) [][]string {
+	lowestPriorityGroup := priorityGroups[len(priorityGroups)-1]
+	if len(lowestPriorityGroup) <= 1 {
+		priorityGroups = priorityGroups[:len(priorityGroups)-1]
+		return priorityGroups
+	}
+
+	for index, elem := range lowestPriorityGroup {
+		if elem == oldestElement {
+			updatedPriorityGroup := append(lowestPriorityGroup[:index], lowestPriorityGroup[index+1:]...)
+			priorityGroups[len(priorityGroups)-1] = updatedPriorityGroup
+			return priorityGroups
+		}
+	}
+	return priorityGroups
+}
+
+func isSyncerPriority(syncer string, priorityGroups [][]string) bool {
+	for _, group := range priorityGroups {
+		for _, bidder := range group {
+			if syncer == bidder {
```

Comment:
Are we comparing syncer keys and bidder names here? If so, there will be a mismatch.